File: src/main/GO/leetcode/lists/topinterview/LeetCode384.go
package topinterview

import "math/rand"

type Solution struct {
	nums   []int
	origin []int
}

func ConstructorNums(nums []int) Solution {
	//fix
	return Solution{nums, append([]int(nil), nums...)}
}

func (this *Solution) Reset() []int {
	return this.origin
}

func (this *Solution) Shuffle() []int {
	for i := 0; i < len(this.nums); i++ {
		r := rand.Intn(len(this.nums) - i)
		this.nums[i], this.nums[r] = this.nums[r], this.nums[i]
	}
	return this.nums
}

File: src/main/java/leetcode/lists/topinterview/LeetCode384.java
package leetcode.lists.topinterview;

import java.util.Arrays;
import java.util.Random;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/30 02:37
 */
public class LeetCode384 {

    class Solution {

        int[] origin;
        int[] arr;

        Random r = new Random();

        public Solution(int[] nums) {
            //fix
            origin = Arrays.copyOf(nums, nums.length);
            arr = nums;
        }

        public int[] reset() {
            return origin;
        }

        public int[] shuffle() {
            for (int i = 0; i < arr.length; i++) {
                //copy
                int index = r.nextInt(arr.length - i);
                int select = arr[index];
                int t = arr[i];
                arr[i] = select;
                arr[index] = t;
            }
            return arr;
        }
    }
}
